hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/RowStatisticsCompactionObserver.java

```diff
@@ -0,0 +1,321 @@
+    this.recorder = recorder;
+    this.tableRecorder = null;
+  }
+
+  public RowStatisticsCompactionObserver() {
+    this(null);
+  }
+
+  @Override
+  public Optional<RegionObserver> getRegionObserver() {
+    return Optional.of(this);
+  }
+
+  @Override
+  public Optional<MasterObserver> getMasterObserver() {
+    return Optional.of(this);
+  }
+
+  @Override
+  public void start(CoprocessorEnvironment e) throws IOException {
+    if (e instanceof RegionCoprocessorEnvironment) {
+      RegionCoprocessorEnvironment regionEnv = (RegionCoprocessorEnvironment) e;
+      if (RowStatisticsUtil.isInternalTable(regionEnv)) {
+        return;
+      }
+
+      String[] configuredBuckets =
+        regionEnv.getConfiguration().getStrings(BlockCacheFactory.BUCKET_CACHE_BUCKETS_KEY);
+      maxCacheSize = DEFAULT_MAX_BUCKET_SIZE;
+      if (configuredBuckets != null && configuredBuckets.length > 0) {
+        String lastBucket = configuredBuckets[configuredBuckets.length - 1];
+        try {
+          maxCacheSize = Integer.parseInt(lastBucket.trim());
+        } catch (NumberFormatException ex) {
+          LOG.warn("Failed to parse {} value {} as int", BlockCacheFactory.BUCKET_CACHE_BUCKETS_KEY,
+            lastBucket, ex);
+        }
+      }
+
+      rowStatisticsDropped =
+        regionEnv.getMetricRegistryForRegionServer().counter(ROW_STATISTICS_DROPPED);
+      rowStatisticsPutFailed =
+        regionEnv.getMetricRegistryForRegionServer().counter(ROW_STATISTICS_PUT_FAILED);
+
+      TableName tableName = regionEnv.getRegionInfo().getTable();
+      TABLE_COUNTERS.merge(tableName.getNameAsString(), 1L, Long::sum);
+    }
+  }
+
+  @Override
+  public void stop(CoprocessorEnvironment e) throws IOException {
+    if (e instanceof RegionCoprocessorEnvironment) {
+      RegionCoprocessorEnvironment regionEnv = (RegionCoprocessorEnvironment) e;
+      if (RowStatisticsUtil.isInternalTable(regionEnv)) {
+        return;
+      }
+      TableName tableName = regionEnv.getRegionInfo().getTable();
+      long tableCount = TABLE_COUNTERS.merge(tableName.getNameAsString(), -1L, Long::sum);
+      if (tableCount == 0) {
+        long regionCount = 0;
+        for (long count : TABLE_COUNTERS.values()) {
+          regionCount += count;
+        }
+        if (regionCount == 0) {
+          regionEnv.getMetricRegistryForRegionServer().remove(ROW_STATISTICS_DROPPED,
+            rowStatisticsDropped);
+          regionEnv.getMetricRegistryForRegionServer().remove(ROW_STATISTICS_PUT_FAILED,
+            rowStatisticsPutFailed);
+          boolean removed = regionEnv.getSharedData().remove(TABLE_RECORDER_KEY, tableRecorder);
+          if (removed) {
+            tableRecorder.close();
+          }
+        }
+      }
+    }
+  }
+
+  @Override
+  public void postStartMaster(ObserverContext<MasterCoprocessorEnvironment> ctx)
+    throws IOException {
+    try (Admin admin = ctx.getEnvironment().getConnection().getAdmin()) {
```

Comment:
Is there a race between master+coprocessor startup and a region server trying to write stats?

Reply:
If the RegionServer tries to record row statistics before the HMasterObserver coprocessor can create the internal table, the [RegionServer will drop the statistics for this region's compaction event](https://github.com/apache/hbase/blob/a32c93444ac6c97971ed29b57b11c5e2f334253b/hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/ringbuffer/RowStatisticsEventHandler.java#L57-L60). The RegionServer will attempt to record the region's updated row statistics on the next one.

**More details:**
- We [wrap the InternalScanner of the preCompact hook with our own RowStatisticsInternalScanner](https://github.com/apache/hbase/blob/a32c93444ac6c97971ed29b57b11c5e2f334253b/hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/RowStatisticsCompactionObserver.java#L221). 
- The [RowStatisticsInternalScanner](https://github.com/apache/hbase/blob/a32c93444ac6c97971ed29b57b11c5e2f334253b/hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/RowStatisticsCompactionObserver.java#L234-L235) records row statistics in the internal row statistics table. It does this by instantiating a new (or retrieving an existing) RowStatisticsTableRecorder object and passing the region's row statistics to it. 
- The RowStatisticsTableRecorder::record method uses a [disruptor to pass row statistics between the data collection and persistence stages of the system](https://github.com/apache/hbase/blob/a32c93444ac6c97971ed29b57b11c5e2f334253b/hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/recorder/RowStatisticsTableRecorder.java#L99-L100).
- In the data persistence layer of the system, [we drop row statistics when we cannot publish them to the internal table](https://github.com/apache/hbase/blob/a32c93444ac6c97971ed29b57b11c5e2f334253b/hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/ringbuffer/RowStatisticsEventHandler.java#L57-L60)

Reply:
👍